plugins/hls-signature-help-plugin/src/Ide/Plugin/SignatureHelp.hs

```diff
@@ -0,0 +1,208 @@
+                            oldPosition
+                            hieAst
+                            getArgumentNumber
+                pure $ Just (functionName, functionType, argumentNumber)
+    case mResult of
+        -- TODO(@linj) what do non-singleton lists mean?
+        Just (functionName : _, functionType : _, argumentNumber : _) -> do
+            pure $ InL $ mkSignatureHelp functionName functionType (fromIntegral argumentNumber - 1)
+        _ -> pure $ InR Null
+
+mkSignatureHelp :: Name -> Text -> UInt -> SignatureHelp
+mkSignatureHelp functionName functionType argumentNumber =
+    let functionNameLabelPrefix = printOutputable (ppr functionName) <> " :: "
+     in SignatureHelp
+            [ SignatureInformation
+                  (functionNameLabelPrefix <> functionType)
+                  Nothing
+                  (Just $ mkArguments (fromIntegral $ T.length functionNameLabelPrefix) functionType)
+                  (Just $ InL argumentNumber)
+            ]
+            (Just 0)
+            (Just $ InL argumentNumber)
+
+-- TODO(@linj) can type string be a multi-line string?
+mkArguments :: UInt -> Text -> [ParameterInformation]
+mkArguments offset functionType =
+    let separator = " -> "
+        separatorLength = fromIntegral $ T.length separator
+        splits = T.breakOnAll separator functionType
+        prefixes = fst <$> splits
+        prefixLengths = fmap (T.length >>> fromIntegral) prefixes
+        ranges =
+            [ ( if previousPrefixLength == 0 then 0 else previousPrefixLength + separatorLength,
+                currentPrefixLength
+              )
+            | (previousPrefixLength, currentPrefixLength) <- zip (0: prefixLengths) prefixLengths
+            ]
+     in [ ParameterInformation (InR range) Nothing
+        | range <- bimap (+offset) (+offset) <$> ranges
+        ]
+
+extractInfoFromSmallestContainingFunctionApplicationAst ::
+    Position -> HieASTs a -> (RealSrcSpan -> HieAST a -> Maybe b) -> [b]
+extractInfoFromSmallestContainingFunctionApplicationAst position hieAsts extractInfo =
+    M.elems $ flip M.mapMaybeWithKey (getAsts hieAsts) $ \hiePath hieAst ->
+        smallestContainingSatisfying (positionToSpan hiePath position) (nodeHasAnnotation ("HsApp", "HsExpr")) hieAst
+            >>= extractInfo (positionToSpan hiePath position)
+    where
+        positionToSpan hiePath position =
+            let loc = mkLoc hiePath position in mkRealSrcSpan loc loc
+        mkLoc (LexicalFastString hiePath) (Position line character) =
+            mkRealSrcLoc hiePath (fromIntegral line + 1) (fromIntegral character + 1)
+
+type Annotation = (FastStringCompat, FastStringCompat)
+
+nodeHasAnnotation :: Annotation -> HieAST a -> Bool
+nodeHasAnnotation annotation = sourceNodeInfo >>> maybe False (isAnnotationInNodeInfo annotation)
+
+-- TODO(@linj): the left most node may not be the function node. example: (if True then f else g) x
```

Comment:
write a test!

Reply:
this is a great example btw. In this case I think we would have to give no signature help, I think, since it's only dynamically known which function we're getting.